README.md
# streamful

Widgets and utilities to deal with streams

## Features

### StreamedLoading

A circular indicator that will show looking to a boolean stream representing the loading status.

You can customize stroke **size** and **color**.

```dart
StreamedLoading(
  stream: _loadingController.stream,
  color: Colors.Blue,
  strokeWidth: 4.0,
);
```

### Loading mixin

A mixin that adds a `BehaviorSubject<bool>` to a class. This way you can easily deal with a stream representing the loading status.

You can set the loading status using `loadStart()` and `loadStop()` functions.
To retrieve the loading status you can use the `isLoading` getter.

### StreamWidget

This is a wrapper of `StreamBuilder`. It allows to define:
- an `onData: (data) {}` callback that must return the Widget to show when the stream has data
- an `onError: (error) {}` callback that must return the Widget to show when the stream has errors
- an onLoad: widget to be returned when the stream has no data nor errors


CHANGELOG.md
# Streamful - Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2021-05-29
### Changed
* Upgrade dependencies [@dariof28](https://github.com/dariof28)
* Migrate to null safety [@dariof28](https://github.com/dariof28)

## [1.0.1] - 2020-12-29
### Changed
* Update dependencies and package description. [@dariof28](https://github.com/dariof28)

## [1.0.0] - 2020-12-29
### Added
* StreamedLoading widget [@dariof28](https://github.com/dariof28)
* Loading mixin [@dariof28](https://github.com/dariof28)
* StreamWidget [@dariof28](https://github.com/dariof28)


test/loading_mixin_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:streamful/src/loading_mixin.dart';

class MyClass with Loading {}

main() {
  test('Loading mixin can start loading', () async {
    var instance = new MyClass();
    instance.loadStart();

    expect(await instance.isLoading.first, true);
  });

  test('Loading mixin can stop loading', () async {
    var instance = new MyClass();
    instance.loadStop();

    expect(await instance.isLoading.first, false);
  });
}


test/stream_widget_test.dart
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:rxdart/rxdart.dart';
import 'package:streamful/streamful.dart';

main() {
  Stream<String> _getStreamData(String value) {
    final controller = BehaviorSubject<String>();
    controller.add(value);
    controller.close();

    return controller.stream;
  }

  Stream<String> _getStreamError(String value) {
    final controller = BehaviorSubject<String>();
    controller.addError(value);
    controller.close();

    return controller.stream;
  }

  Stream<String> _getEmptyStream() {
    final controller = BehaviorSubject<String>();
    controller.close();

    return controller.stream;
  }

  testWidgets('StreamWidget with data returns onData result',
      (WidgetTester tester) async {
    var widget = StreamWidget<String>(
      stream: _getStreamData('success'),
      onData: (data) => Text(data as String, textDirection: TextDirection.ltr),
      onError: (_) => Text('error', textDirection: TextDirection.ltr),
      onLoad: Text('loading...', textDirection: TextDirection.ltr),
    );

    await tester.pumpWidget(widget);
    await tester.pump(Duration.zero);

    expect(find.text('success'), findsOneWidget);
    expect(find.text('error'), findsNothing);
    expect(find.text('loading...'), findsNothing);
  });

  testWidgets('StreamWidget with errors returns onError result',
      (WidgetTester tester) async {
    var widget = StreamWidget<String>(
      stream: _getStreamError('error'),
      onData: (_) => Text('success', textDirection: TextDirection.ltr),
      onError: (object) =>
          Text(object as String, textDirection: TextDirection.ltr),
      onLoad: Text('loading...', textDirection: TextDirection.ltr),
    );

    await tester.pumpWidget(widget);
    await tester.pump(Duration.zero);

    expect(find.text('error'), findsOneWidget);
    expect(find.text('success'), findsNothing);
    expect(find.text('loading...'), findsNothing);
  });

  testWidgets('StreamWidget without data nor errors returns onLoad result',
      (WidgetTester tester) async {
    var widget = StreamWidget<String>(
      stream: _getEmptyStream(),
      onData: (_) => Text('success', textDirection: TextDirection.ltr),
      onError: (_) => Text('error', textDirection: TextDirection.ltr),
      onLoad: Text('loading...', textDirection: TextDirection.ltr),
    );

    await tester.pumpWidget(widget);
    await tester.pump(Duration.zero);

    expect(find.text('loading...'), findsOneWidget);
    expect(find.text('success'), findsNothing);
    expect(find.text('error'), findsNothing);
  });

  testWidgets('StreamWidget with data without onData returns Container',
      (WidgetTester tester) async {
    var widget = StreamWidget<String>(
      stream: _getStreamData('success'),
      onError: (_) => Text('error', textDirection: TextDirection.ltr),
      onLoad: Text('loading...', textDirection: TextDirection.ltr),
    );

    await tester.pumpWidget(widget);
    await tester.pump(Duration.zero);

    expect(find.byType(Container), findsOneWidget);
    expect(find.text('error'), findsNothing);
    expect(find.text('loading...'), findsNothing);
  });

  testWidgets('StreamWidget with errors without onError returns Container',
      (WidgetTester tester) async {
    var widget = StreamWidget<String>(
      stream: _getStreamError('error'),
      onData: (_) => Text('success', textDirection: TextDirection.ltr),
      onLoad: Text('loading...', textDirection: TextDirection.ltr),
    );

    await tester.pumpWidget(widget);
    await tester.pump(Duration.zero);

    expect(find.byType(Container), findsOneWidget);
    expect(find.text('success'), findsNothing);
    expect(find.text('loading...'), findsNothing);
  });

  testWidgets(
      'StreamWidget without data nor errors without onLoad returns Container',
      (WidgetTester tester) async {
    var widget = StreamWidget<String>(
      stream: _getEmptyStream(),
      onData: (_) => Text('success', textDirection: TextDirection.ltr),
      onError: (_) => Text('error', textDirection: TextDirection.ltr),
    );

    await tester.pumpWidget(widget);
    await tester.pump(Duration.zero);

    expect(find.byType(Container), findsOneWidget);
    expect(find.text('success'), findsNothing);
    expect(find.text('error'), findsNothing);
  });
}


test/streamed_loading_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:rxdart/rxdart.dart';
import 'package:streamful/streamful.dart';

main() {
  Stream<bool> _getStream(bool value) {
    final controller = BehaviorSubject<bool>();
    controller.add(value);
    controller.close();

    return controller.stream;
  }

  testWidgets('Display a Container if stream value is false',
      (WidgetTester tester) async {
    await tester.pumpWidget(StreamedLoading(stream: _getStream(false)));
    await tester.pump(Duration.zero);

    expect(find.byType(CircularProgressIndicator), findsNothing);
    expect(find.byType(Container), findsOneWidget);
  });

  testWidgets('Display a CircularProgressIndicator if stream value is true',
      (WidgetTester tester) async {
    await tester.pumpWidget(StreamedLoading(stream: _getStream(true)));
    await tester.pump(Duration.zero);

    expect(find.byType(CircularProgressIndicator), findsOneWidget);
  });

  testWidgets('Can customize CircularProgressIndicator color',
      (WidgetTester tester) async {
    var color = Colors.red;

    await tester.pumpWidget(StreamedLoading(
      stream: _getStream(true),
      color: color,
    ));
    await tester.pump(Duration.zero);

    WidgetPredicate predicate = (Widget widget) =>
        widget is CircularProgressIndicator &&
        widget.valueColor is AlwaysStoppedAnimation &&
        widget.valueColor!.value == color;
    expect(find.byWidgetPredicate(predicate), findsOneWidget);
  });

  testWidgets('Can customize CircularProgressIndicator strokeWidth',
      (WidgetTester tester) async {
    var width = 8.0;

    await tester.pumpWidget(StreamedLoading(
      stream: _getStream(true),
      strokeWidth: width,
    ));
    await tester.pump(Duration.zero);

    WidgetPredicate predicate = (Widget widget) =>
        widget is CircularProgressIndicator && widget.strokeWidth == width;
    expect(find.byWidgetPredicate(predicate), findsOneWidget);
  });

  testWidgets('CircularProgressIndicator strokeWidth is 4.0 by default',
      (WidgetTester tester) async {
    await tester.pumpWidget(StreamedLoading(stream: _getStream(true)));
    await tester.pump(Duration.zero);

    WidgetPredicate predicate = (Widget widget) =>
        widget is CircularProgressIndicator && widget.strokeWidth == 4.0;
    expect(find.byWidgetPredicate(predicate), findsOneWidget);
  });
}


lib/src/stream_widget.dart
import 'package:flutter/material.dart';

typedef DataCallBack<T> = Widget Function(T data);
typedef ErrorCallBack<Object> = Widget Function(Object error);

/// Wrapper of [StreamBuilder].
///
/// You can pass an [onData] callback that will return a Widget that can deal with stream data.
/// You can pass an [onError] callback that will return a Widget that can deal with the stream error.
/// You can pass an [onLoad] widget to be returned while stream has no data nor errors.
///
/// If onData, onError or onLoad are null, a Container is returned for that specif scenario.
class StreamWidget<T> extends StatelessWidget {
  final Stream<T> stream;
  final DataCallBack<T?>? onData;
  final ErrorCallBack<Object?>? onError;
  final Widget? onLoad;

  StreamWidget({required this.stream, this.onData, this.onError, this.onLoad});

  @override
  Widget build(BuildContext context) {
    return StreamBuilder<T>(
      stream: stream,
      builder: (context, snapshot) {
        if (snapshot.hasError) {
          if (onError != null) {
            return onError!(snapshot.error);
          }

          return Container();
        }

        if (snapshot.hasData) {
          if (onData != null) {
            return onData!(snapshot.data);
          }

          return Container();
        }

        return onLoad ?? Container();
      },
    );
  }
}


lib/src/streamed_loading.dart
import 'package:flutter/material.dart';

/// A Circular indicator to be showed when the given [stream] value is true.
class StreamedLoading extends StatelessWidget {
  /// The stream being used as toggle.
  final Stream<bool> stream;

  /// The color of the indicator.
  ///
  /// If null [ThemeData.accentColor] will be used.
  final Color? color;

  /// The width of the line of the indicator.
  final double strokeWidth;

  StreamedLoading({
    Key? key,
    required this.stream,
    this.color,
    this.strokeWidth = 4.0,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return StreamBuilder<bool>(
      stream: stream,
      builder: (BuildContext context, AsyncSnapshot<bool> snapshot) {
        if (snapshot.hasData && snapshot.data!) {
          return Padding(
            padding: const EdgeInsets.all(8.0),
            child: Center(
              child: CircularProgressIndicator(
                valueColor: AlwaysStoppedAnimation(color),
                strokeWidth: strokeWidth,
              ),
            ),
          );
        }

        return Container();
      },
    );
  }
}


example/lib/main.dart
import 'package:example/bloc.dart';
import 'package:flutter/material.dart';
import 'package:streamful/streamful.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Streamful Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Streamful Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  final String? title;

  MyHomePage({Key? key, this.title}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Color? _color = Colors.blue;
  double _strokeWidth = 4.0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(widget.title!)),
      body: Padding(
        padding: EdgeInsets.all(16.0),
        child: _buildContent(),
      ),
    );
  }

  Widget _buildContent() {
    return Column(
      children: [
        Container(
          alignment: Alignment.centerLeft,
          child: Text(
            'Streamful',
            style: TextStyle(fontWeight: FontWeight.bold, fontSize: 22.0),
          ),
        ),
        Padding(
          padding: const EdgeInsets.only(bottom: 8.0),
          child: _buildColorPicker(),
        ),
        Padding(
          padding: const EdgeInsets.only(bottom: 8.0),
          child: _buildStrokeSizer(),
        ),
        Padding(
          padding: const EdgeInsets.only(bottom: 8.0),
          child: _buildActionButtons(),
        ),
        StreamWidget<String?>(
          stream: bloc.data,
          onData: (data) => Text(data!),
          onError: (error) => Text(error.toString()),
          onLoad: StreamedLoading(
            stream: bloc.isLoading,
            color: _color,
            strokeWidth: _strokeWidth,
          ),
        ),
      ],
    );
  }

  Widget _buildColorPicker() {
    return Row(
      children: [
        Expanded(
          child: Padding(
            padding: const EdgeInsets.only(right: 8.0),
            child: Text('Select a color'),
          ),
        ),
        Expanded(
          child: DropdownButton<Color>(
            isExpanded: true,
            value: _color,
            items: _colors,
            onChanged: (color) {
              setState(() {
                _color = color;
              });
            },
          ),
        ),
      ],
    );
  }

  List<DropdownMenuItem<Color>> get _colors {
    Map<String, Color> colors = {
      'Blue': Colors.blue,
      'Red': Colors.red,
      'Orange': Colors.orange,
      'Green': Colors.green,
    };

    List<DropdownMenuItem<Color>> _list = [];

    colors.forEach((key, value) {
      _list.add(
        DropdownMenuItem(child: Text(key), value: value),
      );
    });

    return _list;
  }

  Widget _buildStrokeSizer() {
    return Row(
      children: [
        Expanded(
          child: Padding(
            padding: const EdgeInsets.only(right: 8.0),
            child: Text('Set the stroke width'),
          ),
        ),
        Expanded(
          child: Slider(
            min: 1,
            max: 10,
            value: _strokeWidth,
            onChanged: (value) {
              setState(() {
                _strokeWidth = value;
              });
            },
          ),
        ),
      ],
    );
  }

  Widget _buildActionButtons() {
    return Row(
      children: [
        ElevatedButton(
          onPressed: () {
            bloc.getData();
          },
          child: Text('Success'),
        ),
        Spacer(),
        ElevatedButton(
          onPressed: () {
            bloc.getError();
          },
          child: Text('Error'),
        ),
      ],
    );
  }
}
